Improving ThePlatform Metadata extraction

trainman261 opened the pull request:
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

This improves metadata extraction for sites that use ThePlatform and updates the tests accordingly:

- "standard" metadata fields: tags/keywords and categories.
- A number of metadata fields have site-specific prefixes, but many of those metadata fields themselves are not site-specific (for instance creator, series, season number in this PR). I had started a discussion on this topic on Discord a while ago, but the discussion stalled IMO without a resolution. Consider this a proposal continuing the discussion that is open to changes. The proposed method could also help solve an issue in the NBC IE where it subclasses ThePlatformIE instead of subclassing InfoExtractor which AFAIK is not best practice (see the comment on line 32).

I have done the same level of regression testing as for my previous ThePlatformIE PR and have fixed all regression issues that appeared.

I am also now proposing to add the media_type field with this PR. This should be more carefully looked at, since I don't know much about the code outside of the extractor classes - I've tried to find where changes are needed but I only found the explanation in the common extractor class. Please add any other changes that may be necessary.

Fixes #


<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

bashonly commented on yt_dlp/extractor/theplatform.py:
```diff
@@ -103,6 +105,36 @@ def _add_chapter(start_time, end_time):
+        tags = []
+        if (info_keywords_str is not None) and (info_keywords_str != ''):
+            tags = re.split(', |; |,', info_keywords_str)
+        location = None
+        series = None
+        season_number = None
```
these should be moved down by where the values are extracted, but my other suggestion below will make initializing the variables unnecessary
```suggestion
```

trainman261 replied:
Yes, that other proposition is better in total.

bashonly commented on yt_dlp/extractor/theplatform.py:
```diff
@@ -103,6 +105,36 @@ def _add_chapter(start_time, end_time):
+                if (x.get('label') is None) or (x.get('label') == 'category'):
+                    categories.append(x.get('name'))
+
+        # A number of sites have a prefix in front of some info keys followed by a '$' symbol.
+        # Search for known keys with the prefix.
+        for key in info.keys():
+            if re.match('.*\\$region', key):
+                location = info[key]
+            if re.match('.*\\$show', key):
+                series = info[key]
+            if re.match('.*\\$seasonNumber', key):
+                season_number = str_to_int(info[key])
```
With regex patterns, always use a raw string, e.g. `r'Backslashes\sAre\sOK'`. but regex is unnecessary anyways

We can use a helper function to reduce code duplication, except instead of using vars for `location` etc we could just inline them in the return below:
```suggestion
        def extract_site_specific_field(field):
            # A number of sites have custom-prefixed keys, e.g. 'cbc$seasonNumber'
            return next((info[k] for k in info if k.endswith(f'${field}')), None)
        
        location = extract_site_specific_field('region')
        series = extract_site_specific_field('show')
        season_number = int_or_none(extract_site_specific_field('seasonNumber'))
```

bashonly commented on yt_dlp/extractor/theplatform.py:
```diff
@@ -103,6 +105,36 @@ def _add_chapter(start_time, end_time):
+                series = info[key]
+            if re.match('.*\\$seasonNumber', key):
+                season_number = str_to_int(info[key])
+            # the following can be uncommented as soon as #7838 is merged:
+            # if (re.match('.*\programmingType', key)) or (re.match('.*\type', key)):
+                # media_type = info[key]
```
#7838 was merged already without adding the `media_type` field -- feel free to propose its addition in this PR

trainman261 replied:
I've added it - see the description and my comment below.

bashonly commented on yt_dlp/extractor/theplatform.py:
```diff
@@ -113,6 +145,14 @@ def _add_chapter(start_time, end_time):
             'timestamp': int_or_none(info.get('pubDate'), 1000) or None,
             'uploader': info.get('billingCode'),
             'chapters': chapters,
+            'creator': info.get('author', {str_or_none}) if info.get('author', {str_or_none}) != '' else None,
```
this should suffice
```suggestion
            'creator': traverse_obj(info, ('author', {str})) or None,
```

bashonly commented on yt_dlp/extractor/theplatform.py:
```diff
@@ -103,6 +105,36 @@ def _add_chapter(start_time, end_time):
+        season_number = None
+        # The following can be uncommented as soon as #7838 is merged:
+        # media_type = None
+        categories = []
+        categories_data = info.get('categories') or []
+        for x in categories_data:
+            if x.get('name') is not None:
+                # Sometimes, there will be several kinds of categories
+                # in this case, it will have a label field with a value 'category'
+                if (x.get('label') is None) or (x.get('label') == 'category'):
+                    categories.append(x.get('name'))
```
we can use traversal filtering to simplify this part
```suggestion
        categories = traverse_obj(info, (
            'categories', lambda _, v: v.get('label') in ('category', None), 'name', {str})) or None
```
we can inline this in the return too

bashonly commented:
And, unrelated to my review:

> The proposed method could also help solve an issue in the NBC IE where it subclasses ThePlatformIE instead of subclassing InfoExtractor 

I am thinking these multi-purpose extractors such as `ThePlatformIE` and `AnvatoIE` etc should be refactored so that all of their useful methods are in a baseclass, and so that the many extractors improperly subclassing from concrete IEs can properly subclass from the refactored baseclasses instead

trainman261 commented on yt_dlp/extractor/theplatform.py:
```diff
@@ -103,6 +105,36 @@ def _add_chapter(start_time, end_time):
             for chapter in tp_chapters[:-1]:
                 _add_chapter(chapter.get('startTime'), chapter.get('endTime'))
             _add_chapter(tp_chapters[-1].get('startTime'), tp_chapters[-1].get('endTime') or duration)
+        info_keywords_str = info.get('keywords', {str_or_none})
+        tags = []
+        if (info_keywords_str is not None) and (info_keywords_str != ''):
+            tags = re.split(', |; |,', info_keywords_str)
```
Yes, that's it. I've moved it into the return and will push it shortly.

seproDev commented on yt_dlp/extractor/common.py:
```diff
@@ -379,6 +379,8 @@ class InfoExtractor:
                     'private', 'premium_only', 'subscriber_only', 'needs_auth',
                     'unlisted' or 'public'. Use 'InfoExtractor._availability'
                     to set it
+    media_type:     The type of media, for instance a full show, an excerpt, a highlight, a trailer
+                    or something similar that classifies what kind of video or audio this is.
```
`media_type` will also need to be documented in the readme

trainman261 replied:
I've just added a commit, where I've shortened the description to match the others

seproDev commented on yt_dlp/extractor/cbc.py:
```diff
@@ -180,6 +180,24 @@ class CBCPlayerIE(InfoExtractor):
             'thumbnail': 'http://thumbnails.cbc.ca/maven_legacy/thumbnails/sonali-karnick-220.jpg',
             'chapters': [],
             'duration': 494.811,
+            'categories': [
+                'AudioMobile/All in a Weekend Montreal',
+            ],
+            'tags': [
+                'CBC',
+                'CBC Radio',
+                'Montreal',
+                'Quebec',
+                'All In A Weekend',
+                'Sonali Karnick',
+                'hackathon',
+                'listenmontreal',
+            ],
```
```suggestion
            'categories': ['AudioMobile/All in a Weekend Montreal'],
            'tags': 'count:8',
```
Could you please make these changes in the other instances as well?

trainman261 replied:
Will be adding the changes to the other tests in a bit. Are tests not supposed to be that exact?

seproDev replied:
In general, tags/categories tests should be formatted to stay in one line. If there are too many to list them in one line, I prefer to use count instead.

trainman261 replied:
Ah crap, now I've changed a number of tests to `count:1`. The tests have run through successfully, I'll change it back and re-run the tests tomorrow, been a long day today.
But thanks for the explanation, now I understand.

trainman261 replied:
OK, I've corrected it back.

seproDev commented on yt_dlp/extractor/mediaset.py:
```diff
@@ -73,6 +73,10 @@ class MediasetIE(ThePlatformBaseIE):
             'season_number': 5,
             'episode_number': 5,
             'chapters': [{'start_time': 0.0, 'end_time': 3409.08}, {'start_time': 3409.08, 'end_time': 6565.008}],
+            'creator': None,
```
None values should not be added to the test
```suggestion
```

seproDev commented on yt_dlp/extractor/theplatform.py:
```diff
@@ -113,6 +121,14 @@ def _add_chapter(start_time, end_time):
+            'location': extract_site_specific_field('region'),
+            'series': extract_site_specific_field('show'),
+            'season_number': int_or_none(extract_site_specific_field('seasonNumber')),
+            'media_type': info_media_type,
```
Can be inlined
```suggestion
            'media_type': extract_site_specific_field('programmingType') or extract_site_specific_field('type'),
```

seproDev commented on yt_dlp/extractor/theplatform.py:
```diff
@@ -104,6 +104,14 @@ def _add_chapter(start_time, end_time):
+            # A number of sites have custom-prefixed keys, e.g. 'cbc$seasonNumber'
+            return next((info[k] for k in info if k.endswith(f'${field}') and info[k] != ''), None)
+
+        info_media_type = extract_site_specific_field('programmingType')
+        if not info_media_type:
+            info_media_type = extract_site_specific_field('type')
+
```
```suggestion
```

seproDev commented on yt_dlp/extractor/cbc.py:
```diff
@@ -195,25 +202,41 @@ class CBCPlayerIE(InfoExtractor):
             'chapters': [],
             'duration': 186.867,
+            'series': 'CBC News: Windsor at 6:00',
+            'categories': [
+                'News/Canada/Windsor',
+            ],
+            'location': 'Windsor',
+            'tags': [
+                'cancer',
+            ],
```
```suggestion
            'categories': ['News/Canada/Windsor'],
            'location': 'Windsor',
            'tags': ['cancer'],
```

seproDev commented on yt_dlp/extractor/mediaset.py:
```diff
@@ -73,6 +73,9 @@ class MediasetIE(ThePlatformBaseIE):
             'season_number': 5,
             'episode_number': 5,
             'chapters': [{'start_time': 0.0, 'end_time': 3409.08}, {'start_time': 3409.08, 'end_time': 6565.008}],
+            'categories': [
+                'Informazione',
+            ],
```
```suggestion
            'categories': ['Informazione'],
```

seproDev commented on yt_dlp/extractor/mediaset.py:
```diff
@@ -149,6 +152,9 @@ class MediasetIE(ThePlatformBaseIE):
             'season_number': 12,
             'episode': 'Episode 8',
             'episode_number': 8,
+            'categories': [
+                'Intrattenimento',
+            ]
```
```suggestion
            'categories': ['Intrattenimento'],
```

seproDev commented on yt_dlp/extractor/nbc.py:
```diff
@@ -53,6 +53,11 @@ class NBCIE(ThePlatformIE):  # XXX: Do not subclass from concrete IE
                 'chapters': 'count:1',
                 'tags': 'count:4',
                 'thumbnail': r're:https?://.+\.jpg',
+                'categories': [
+                    'Series/The Tonight Show Starring Jimmy Fallon'
+                ],
```
```suggestion
                'categories': ['Series/The Tonight Show Starring Jimmy Fallon'],
```

seproDev commented on yt_dlp/extractor/nbc.py:
```diff
@@ -131,6 +136,10 @@ class NBCIE(ThePlatformIE):  # XXX: Do not subclass from concrete IE
                 'tags': 'count:10',
                 'age_limit': 0,
                 'thumbnail': r're:https?://.+\.jpg',
+                'categories': [
+                    'Series/Quantum Leap 2022',
+                ],
```
```suggestion
                'categories': ['Series/Quantum Leap 2022'],
```

seproDev commented on yt_dlp/extractor/nbc.py:
```diff
@@ -53,6 +53,9 @@ class NBCIE(ThePlatformIE):  # XXX: Do not subclass from concrete IE
                 'tags': 'count:4',
                 'thumbnail': r're:https?://.+\.jpg',
+                'categories': ['Series/The Tonight Show Starring Jimmy Fallon'],
+                'creator': None,
```
```suggestion
```

seproDev commented:
I just noticed that the `CWTVIE` and `AENetworksIE` test also needs to be updated

pukkandan commented on yt_dlp/extractor/theplatform.py:
```diff
@@ -104,6 +104,10 @@ def _add_chapter(start_time, end_time):
 
+        def extract_site_specific_field(field):
+            # A number of sites have custom-prefixed keys, e.g. 'cbc$seasonNumber'
+            return next((info[k] for k in info if k.endswith(f'${field}') and info[k] != ''), None)
```
```suggestion
            return traverse_obj(info, lambda k, v: v and k.endswith(f'${field}', get_all=False)
```

trainman261 commented:
@seproDev I've updated everything I can - unfortunately, those tests are either geo-restricted, subscriber-only or broken. If anyone has access they can feel free to add them.

seproDev commented:
I've pushed those changes myself

trainman261 commented:
I've merged everything together and addressed all the issues. Anything else?

@pukkandan I've noticed Pypy crashing on my system occasionally as well. Is there any pattern that could be reported as an error (to Pypy devs)?
